### PR TITLE
Implement `conj`, `conj!` for `TracedRArray`

### LINKED ISSUE #168: Infinte recursion when compiling a function
CC @mofeing 
I have been trying to compile a function that computes a contract operation given some parameters. Nevertheless, I got a strange `StackOverflowError`. Context:
```julia
using Yao
using Enzyme
using Reactant
using Tenet
using EinExprs
using Adapt
# Define the ansatz circuit
function ansatz(params)
    chain(2,
        put(1=>Ry(params[1])),
        put(2=>Ry(params[2])),
        # control(1, 2=>Ry(params[3])),
        put(1=>Ry(params[4])),
        put(2=>Ry(params[5]))
    )
end

ansatz_circ = ansatz(rand(5))

# Function to compute the expectation value
function expectation(params)
    H = chain(2,
        put(1=>Z),
        put(2=>Z),
        # control(1, 2=>X)
    )
    circuit = ansatz(params)

    qcirc = Tenet.Quantum(circuit)
    H = Tenet.Quantum(H)

    psi000 = Tenet.Quantum(Tenet.Product(fill([1, 0], 2))) #|00>
    state = merge(psi000, qcirc) # circuit|00>

    expval = merge(state, H, state')
    return expval
end

qtn = expectation(rand(5))
tn = Tenet.TensorNetwork(expectation(rand(5)))
path = einexpr(tn; optimizer=Greedy())

function loss(params)
    tn = Tenet.TensorNetwork(expectation(params))
    # path = einexpr(tn; optimizer=Greedy())
    return Tenet.contract(tn; path)
end

params = rand(5)
params′ = adapt(ConcreteRArray, params)
```

Then:
```julia

julia> f = @compile loss(params′)
┌ Warning: Performing scalar indexing on task Task (runnable) @0x000071040fa00010.
│ Invocation resulted in scalar indexing of a TracedRArray.
│ This is typically caused by calling an iterating implementation of a method.
│ Such implementations *do not* execute on device, but very slowly on the CPU,
│ and require expensive copies and synchronization each time and therefore should be avoided.
└ @ Reactant ~/.julia/packages/Reactant/9aeyF/src/TracedRArray.jl:53
┌ Warning: Performing scalar indexing on task Task (runnable) @0x000071040fa00010.
│ Invocation resulted in scalar indexing of a TracedRArray.
│ This is typically caused by calling an iterating implementation of a method.
│ Such implementations *do not* execute on device, but very slowly on the CPU,
│ and require expensive copies and synchronization each time and therefore should be avoided.
└ @ Reactant ~/.julia/packages/Reactant/9aeyF/src/TracedRArray.jl:53
┌ Warning: Performing scalar indexing on task Task (runnable) @0x000071040fa00010.
│ Invocation resulted in scalar indexing of a TracedRArray.
│ This is typically caused by calling an iterating implementation of a method.
│ Such implementations *do not* execute on device, but very slowly on the CPU,
│ and require expensive copies and synchronization each time and therefore should be avoided.
└ @ Reactant ~/.julia/packages/Reactant/9aeyF/src/TracedRArray.jl:53
┌ Warning: Performing scalar indexing on task Task (runnable) @0x000071040fa00010.
│ Invocation resulted in scalar indexing of a TracedRArray.
│ This is typically caused by calling an iterating implementation of a method.
│ Such implementations *do not* execute on device, but very slowly on the CPU,
│ and require expensive copies and synchronization each time and therefore should be avoided.
└ @ Reactant ~/.julia/packages/Reactant/9aeyF/src/TracedRArray.jl:53
┌ Warning: Performing scalar indexing on task Task (runnable) @0x000071040fa00010.
│ Invocation resulted in scalar indexing of a TracedRArray.
│ This is typically caused by calling an iterating implementation of a method.
│ Such implementations *do not* execute on device, but very slowly on the CPU,
│ and require expensive copies and synchronization each time and therefore should be avoided.
└ @ Reactant ~/.julia/packages/Reactant/9aeyF/src/TracedRArray.jl:53
ERROR: StackOverflowError:
Stacktrace:
     [1] mlirOperationCreate
       @ ~/.julia/packages/Reactant/9aeyF/src/mlir/libMLIR_h.jl:985 [inlined]
     [2] create_operation(name::String, loc::Reactant.MLIR.IR.Location; results::Vector{…}, operands::Vector{…}, owned_regions::Vector{…}, successors::Vector{…}, attributes::Vector{…}, result_inference::Bool)
       @ Reactant.MLIR.IR ~/.julia/packages/Reactant/9aeyF/src/mlir/IR/Operation.jl:315
     [3] create_operation
       @ ~/.julia/packages/Reactant/9aeyF/src/mlir/IR/Operation.jl:273 [inlined]
     [4] broadcast_in_dim(operand::Reactant.MLIR.IR.Value; result_0::Reactant.MLIR.IR.Type, broadcast_dimensions::Reactant.MLIR.IR.Attribute, location::Reactant.MLIR.IR.Location)
       @ Reactant.MLIR.Dialects.stablehlo ~/.julia/artifacts/6c9478736605d4915b52b83803cbe351dc32bd48/StableHLO.inc.jl:430
     [5] broadcast_in_dim
       @ ~/.julia/artifacts/6c9478736605d4915b52b83803cbe351dc32bd48/StableHLO.inc.jl:423 [inlined]
     [6] broadcast_to_size_internal(x::Reactant.TracedRArray{ComplexF64, 0}, rsize::Base.Generator{Base.Iterators.Zip{Tuple{Tuple{}, Tuple{}}}, Reactant.var"#96#97"})
       @ Reactant ~/.julia/packages/Reactant/9aeyF/src/TracedRArray.jl:737
     [7] broadcast_to_size
       @ ~/.julia/packages/Reactant/9aeyF/src/TracedRArray.jl:714 [inlined]
     [8] broadcast_to_size
       @ ~/.julia/packages/Reactant/9aeyF/src/TracedRArray.jl:719 [inlined]
     [9] #98
       @ ./none:0 [inlined]
    [10] iterate
       @ ./generator.jl:47 [inlined]
    [11] _copyto!(dest::Reactant.TracedRArray{ComplexF64, 0}, bc::Base.Broadcast.Broadcasted{Nothing, Nothing, typeof(conj), Tuple{Reactant.TracedRArray{ComplexF64, 0}}})
       @ Reactant ~/.julia/packages/Reactant/9aeyF/src/TracedRArray.jl:759
    [12] copyto!(dest::Reactant.TracedRArray{ComplexF64, 0}, bc::Base.Broadcast.Broadcasted{Nothing, Nothing, typeof(conj), Tuple{Reactant.TracedRArray{ComplexF64, 0}}})
       @ Reactant ~/.julia/packages/Reactant/9aeyF/src/TracedRArray.jl:681
    [13] copyto!(dest::Reactant.TracedRArray{ComplexF64, 0}, bc::Base.Broadcast.Broadcasted{Reactant.AbstractReactantArrayStyle{0}, Nothing, typeof(conj), Tuple{Reactant.TracedRArray{ComplexF64, 0}}})
       @ Base.Broadcast ./broadcast.jl:967
    [14] copy(bc::Base.Broadcast.Broadcasted{Reactant.AbstractReactantArrayStyle{0}, Nothing, typeof(conj), Tuple{Reactant.TracedRArray{ComplexF64, 0}}})
       @ Reactant ~/.julia/packages/Reactant/9aeyF/src/TracedRArray.jl:641
    [15] materialize
       @ ./broadcast.jl:903 [inlined]
    [16] broadcast_preserving_zero_d(f::Function, As::Reactant.TracedRArray{ComplexF64, 0})
       @ Base.Broadcast ./broadcast.jl:892
    [17] conj(A::Reactant.TracedRArray{ComplexF64, 0})
       @ Base ./abstractarraymath.jl:145
    [18] elem_apply(f::Function, args::Reactant.TracedRArray{ComplexF64, 0})
       @ Reactant ~/.julia/packages/Reactant/9aeyF/src/TracedRArray.jl:354
--- the last 8 lines are repeated 6663 more times ---
 [53323] _copyto!(dest::Reactant.TracedRArray{ComplexF64, 0}, bc::Base.Broadcast.Broadcasted{Nothing, Nothing, typeof(conj), Tuple{Reactant.TracedRArray{ComplexF64, 0}}})
       @ Reactant ~/.julia/packages/Reactant/9aeyF/src/TracedRArray.jl:759
 [53324] copyto!(dest::Reactant.TracedRArray{ComplexF64, 0}, bc::Base.Broadcast.Broadcasted{Nothing, Nothing, typeof(conj), Tuple{Reactant.TracedRArray{ComplexF64, 0}}})
       @ Reactant ~/.julia/packages/Reactant/9aeyF/src/TracedRArray.jl:681
 [53325] copyto!(dest::Reactant.TracedRArray{ComplexF64, 0}, bc::Base.Broadcast.Broadcasted{Reactant.AbstractReactantArrayStyle{0}, Nothing, typeof(conj), Tuple{Reactant.TracedRArray{ComplexF64, 0}}})
       @ Base.Broadcast ./broadcast.jl:967
 [53326] copy(bc::Base.Broadcast.Broadcasted{Reactant.AbstractReactantArrayStyle{0}, Nothing, typeof(conj), Tuple{Reactant.TracedRArray{ComplexF64, 0}}})
       @ Reactant ~/.julia/packages/Reactant/9aeyF/src/TracedRArray.jl:641
 [53327] materialize
       @ ./broadcast.jl:903 [inlined]
 [53328] broadcast_preserving_zero_d(f::Function, As::Reactant.TracedRArray{ComplexF64, 0})
       @ Base.Broadcast ./broadcast.jl:892
Some type information was truncated. Use `show(err)` to see complete types.
```

Do know what may have caused this?

Thank you!

### PR DESCRIPTION
**Pull request by mofeing**
Fixes #168

# To do

- [x] Implement `conj` for `TracedRNumber`
- [x] Implement in-place `conj!`
- [x] Write tests for `conj`, `conj!`
- [x] Implement `real`, `imag`
- [x] Write tests for `real`, `imag`
- [x] Implement adjoints of `real`, `imag` in Enzyme-JAX https://github.com/EnzymeAD/Enzyme-JAX/pull/145
  - [ ] update Enzyme-JAX commit on ReactantExtra

### CODE REVIEW

**Comment by mofeing** [comment]
`promote_to` on a `Array{ComplexF64}` was failing because `DenseElementsAttribute` was not implemented for it. I implemented but now segfaults. Here is the stacktrace:

```
Process 45118 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x3ff0000000000000)
    frame #0: 0x0000000355f2a5e8 libReactantExtra.dylib`mlir::AttributeStorage::getAbstractAttribute() const + 24
libReactantExtra.dylib`mlir::AttributeStorage::getAbstractAttribute:
->  0x355f2a5e8 <+24>: ldr    x8, [x8]
    0x355f2a5ec <+28>: mov    w9, #0x0 ; =0 
    0x355f2a5f0 <+32>: stur   w9, [x29, #-0xc]
    0x355f2a5f4 <+36>: cbz    x8, 0x355f2a608 ; <+56>
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x3ff0000000000000)
  * frame #0: 0x0000000355f2a5e8 libReactantExtra.dylib`mlir::AttributeStorage::getAbstractAttribute() const + 24
    frame #1: 0x0000000355f2a57c libReactantExtra.dylib`mlir::Attribute::getTypeID() + 28
    frame #2: 0x00000003562ad4fc libReactantExtra.dylib`bool mlir::detail::StorageUserBase<mlir::ArrayAttr, mlir::Attribute, mlir::detail::ArrayAttrStorage, mlir::detail::AttributeUniquer>::classof<mlir::Attribute>(mlir::Attribute) + 28
    frame #3: 0x00000003562adcb0 libReactantExtra.dylib`llvm::CastInfo<mlir::ArrayAttr, mlir::Attribute const, void>::isPossible(mlir::Attribute) + 32
    frame #4: 0x00000003562adbb4 libReactantExtra.dylib`bool llvm::isa<mlir::ArrayAttr, mlir::Attribute>(mlir::Attribute const&) + 36
    frame #5: 0x0000000363dd6270 libReactantExtra.dylib`mlir::DenseElementsAttr::get(mlir::ShapedType, llvm::ArrayRef<mlir::Attribute>) + 1104
    frame #6: 0x0000000363697218 libReactantExtra.dylib`mlirDenseElementsAttrGet + 184
    frame #7: 0x000000014697c0c0
    frame #8: 0x000000013a3646c0
    frame #9: 0x000000012ed2c928
    frame #10: 0x000000010ee6412c
    frame #11: 0x0000000102f0c1c8
    frame #12: 0x0000000102f1c060
    frame #13: 0x0000000100a325c8 libjulia-internal.1.10.5.dylib`do_call [inlined] jl_apply(args=<unavailable>, nargs=<unavailable>) at julia.h:1982:12 [opt]
    frame #14: 0x0000000100a325b8 libjulia-internal.1.10.5.dylib`do_call(args=<unavailable>, nargs=<unavailable>, s=<unavailable>) at interpreter.c:126:26 [opt]
    frame #15: 0x0000000100a30dc4 libjulia-internal.1.10.5.dylib`eval_body(stmts=0x000000010948c3d0, s=0x000000016fdfb800, ip=9, toplevel=<unavailable>) at interpreter.c:0 [opt]
    frame #16: 0x0000000100a313e4 libjulia-internal.1.10.5.dylib`jl_interpret_toplevel_thunk(m=0x0000000141384ad0, src=<unavailable>) at interpreter.c:775:21 [opt]
    frame #17: 0x0000000100a49928 libjulia-internal.1.10.5.dylib`jl_toplevel_eval_flex(m=0x0000000141384ad0, e=0x000000010b235b50, fast=1, expanded=<unavailable>) at toplevel.c:934:18 [opt]
    frame #18: 0x0000000100a496f0 libjulia-internal.1.10.5.dylib`jl_toplevel_eval_flex(m=0x0000000141384ad0, e=0x000000010b235cf0, fast=1, expanded=<unavailable>) at toplevel.c:877:19 [opt]
    frame #19: 0x0000000100a4a5d8 libjulia-internal.1.10.5.dylib`ijl_toplevel_eval_in [inlined] ijl_toplevel_eval(m=0x0000000141384ad0, v=0x000000010b235cf0) at toplevel.c:943:12 [opt]
    frame #20: 0x0000000100a4a5cc libjulia-internal.1.10.5.dylib`ijl_toplevel_eval_in(m=0x0000000141384ad0, ex=0x000000010b235cf0) at toplevel.c:985:13 [opt]
    frame #21: 0x000000013cf8bbf4 sys.dylib`japi1_include_string_81176.1 at boot.jl:385
    frame #22: 0x0000000100a17eac libjulia-internal.1.10.5.dylib`ijl_apply_generic [inlined] _jl_invoke(F=0x0000000140644330, args=0x000000016fdfc0b0, nargs=4, mfunc=0x00000001406450b0, world=<unavailable>) at gf.c:0 [opt]
    frame #23: 0x0000000100a17e40 libjulia-internal.1.10.5.dylib`ijl_apply_generic(F=0x0000000140644330, args=0x000000016fdfc0b0, nargs=<unavailable>) at gf.c:3077:12 [opt]
    frame #24: 0x000000013cf7ae2c sys.dylib`japi1__include_81184.1 at loading.jl:2136
    frame #25: 0x000000013c9d7928 sys.dylib`julia_include_46590.1 at Base.jl:495
    frame #26: 0x000000013ced2138 sys.dylib`jfptr_include_46591.1 + 60
    frame #27: 0x0000000100a17eac libjulia-internal.1.10.5.dylib`ijl_apply_generic [inlined] _jl_invoke(F=0x000000013d775570, args=0x000000016fdfdcd0, nargs=2, mfunc=0x000000013d775dd0, world=<unavailable>) at gf.c:0 [opt]
    frame #28: 0x0000000100a17e40 libjulia-internal.1.10.5.dylib`ijl_apply_generic(F=0x000000013d775570, args=0x000000016fdfdcd0, nargs=<unavailable>) at gf.c:3077:12 [opt]
    frame #29: 0x000000013bf510ec sys.dylib`julia_exec_options_82788.1 at client.jl:318
    frame #30: 0x000000013cdceaac sys.dylib`julia__start_82926.1 at client.jl:552
    frame #31: 0x000000013cdcebd4 sys.dylib`jfptr__start_82927.1 + 36
    frame #32: 0x0000000100a17eac libjulia-internal.1.10.5.dylib`ijl_apply_generic [inlined] _jl_invoke(F=0x00000001401e74b0, args=0x000000016fdfe1f8, nargs=0, mfunc=0x00000001401e7340, world=<unavailable>) at gf.c:0 [opt]
    frame #33: 0x0000000100a17e40 libjulia-internal.1.10.5.dylib`ijl_apply_generic(F=0x00000001401e74b0, args=0x000000016fdfe1f8, nargs=<unavailable>) at gf.c:3077:12 [opt]
    frame #34: 0x0000000100a7590c libjulia-internal.1.10.5.dylib`true_main [inlined] jl_apply(args=0x000000016fdfe1f0, nargs=1) at julia.h:1982:12 [opt]
    frame #35: 0x0000000100a758f8 libjulia-internal.1.10.5.dylib`true_main(argc=<unavailable>, argv=<unavailable>) at jlapi.c:582:13 [opt]
    frame #36: 0x0000000100a75800 libjulia-internal.1.10.5.dylib`jl_repl_entrypoint(argc=<unavailable>, argv=<unavailable>) at jlapi.c:731:15 [opt]
    frame #37: 0x0000000100003f6c julia`main + 12
    frame #38: 0x000000019589a0e0 dyld`start + 2360
```

**Comment by mofeing** [comment]
@Pangoraw any news on the adjoints of `stablehlo.real`, `stablehlo.imag`?

File: src/mlir/IR/Attribute.jl [hunk 1]
```diff
@@ -564,6 +564,14 @@ function DenseElementsAttribute(values::AbstractArray{Float16})
     )
 end
 
+function DenseElementsAttribute(values::AbstractArray{Complex{T}}) where {T}
+    # TODO we might need to transpose because MLIR expects row-major order
+    shaped_type = TensorType(size(values), Type(Complex{T}))
+    return Attribute(
+        API.mlirDenseElementsAttrRawBufferGet(shaped_type, length(values), pointer(values))
```
[the remark this answers is not in the document]
Comment (mofeing) [reply src/mlir/IR/Attribute.jl:571]: mmm makes sense! that would explain some of the issues I'm having with complex numbers

[the remark this answers is not in the document]
Comment (mofeing) [reply src/mlir/IR/Attribute.jl:571]: for the row-major order... isn't the transpose automatically applied in some pass?
the transpose is not explicitly call on these examples:
- https://github.com/EnzymeAD/Reactant.jl/blob/4a3e671fc2774c71a1dfc55dd655895be3580180/src/TracedRArray.jl#L594-L595
- https://github.com/EnzymeAD/Reactant.jl/blob/4a3e671fc2774c71a1dfc55dd655895be3580180/src/TracedRArray.jl#L267-L268

[the remark this answers is not in the document]
Comment (Pangoraw) [reply src/mlir/IR/Attribute.jl]: yeah so maybe those are wrong too (in MLIR.jl too). I personally don't recall ever making an attribute for N-dim arrays

[the remark this answers is not in the document]
Comment (mofeing) [reply src/mlir/IR/Attribute.jl]: it's sth we can easily check with numerical simulations once we have the StableHLO interpreter

(we should also add numerical tests in MLIR.jl, but I will invest into this once we use upstream MLIR.jl in Reactant)

meanwhile let's leave it like the rest of things (which seems to work) and fix it all at once if we it's true that the order of indices is wrong

File: src/mlir/IR/Type.jl [hunk 2]
```diff
@@ -85,13 +85,6 @@ Creates a signless integer type of the given bitwidth in the context. The type i
 Type(T::Core.Type{<:Integer}; context::Context=context()) =
     Type(API.mlirIntegerTypeGet(context, sizeof(T) * 8))
 
-"""
-    Type(T::Core.Type{<:Complex}; context=context())
-
-Creates a complex type with the given element type.
-"""
-Type(::Core.Type{Complex{T}}; context=context()) where {T} = Type(API.mlirComplexTypeGet(Type(T)))
-
```
Comment (mofeing) [inline src/mlir/IR/Type.jl:94]: @Pangoraw this was already implemented, but I saw you reimplemented it in #195? It was giving me problems in precompilation

**Comment by mofeing** [comment]
This is weird

```
Expression: only(f(x_concrete)) == imag(x)
   Evaluated: 1.0 == 0.0
```

It seems like `stablehlo.imag` returns the inputs when it's a complex number/tensor? Something looks weird there.

File: src/TracedRNumber.jl [hunk 3]
```diff
@@ -199,6 +199,36 @@ for (jlop, hloop) in (
...
+function Base.real(x::TracedRNumber{Complex{T}}) where {T}
+    return TracedRNumber{T}(
+        (),
+        MLIR.IR.result(
+            MLIR.Dialects.stablehlo.real(x.mlir_data; result=mlir_type(TracedRNumber{T})), 1
+        ),
+    )
+end
+
+Base.imag(x::TracedRNumber) = x
```
Comment (Pangoraw) [inline src/TracedRNumber.jl:222]: that's your problem

```suggestion
Base.imag(x::TracedRNumber) = zero(x)
```

File: src/mlir/IR/Type.jl (hunk 2 above)
Comment (Pangoraw) [reply src/mlir/IR/Type.jl:94]: ah right, those are essentially the same things sorry

File: src/TracedRNumber.jl (hunk 3 above)
Comment (wsmoses) [reply src/TracedRNumber.jl:222]: I mean tracedrnumber could be complex as well, no?

Comment (Pangoraw) [reply src/TracedRNumber.jl:222]: That should be covered by the next method:

https://github.com/EnzymeAD/Reactant.jl/blob/ba4dc4c4e5f389049e741c7bf642163ed26c0609/src/TracedRNumber.jl#L223-L230

Comment (mofeing) [reply src/TracedRNumber.jl:222]: Yeah, I meant is weird from StableHLO guys to define `imag` on reals that way.

Comment (mofeing) [reply src/TracedRNumber.jl:222]: Oh, so according to the StableHLO, it should return 0 https://openxla.org/stablehlo/spec#imag

So... it's a bug in XLA?

**Comment by mofeing** [comment]
The bug is ours

```julia
julia> @code_hlo conj(x_concrete)
Module:
module attributes {transform.with_named_sequence} {
  func.func @main(%arg0: tensor<i64>) {
    return
  }
}

julia> @code_hlo optimize=false conj(x_concrete)
Module:
module {
  func.func @main(%arg0: tensor<i64>) -> tensor<i64> {
    %0 = stablehlo.transpose %arg0, dims = [] : (tensor<i64>) -> tensor<i64>
    %1 = stablehlo.transpose %0, dims = [] : (tensor<i64>) -> tensor<i64>
    return %1 : tensor<i64>
  }
}
```

File: src/TracedRNumber.jl (hunk 3 above)
Comment (wsmoses) [reply src/TracedRNumber.jl:222]: why not just always emit the imag op. Presumably we can optimize it out if it's not complex

**Pull request merged**